pad_server/src/controller.c

```diff
@@ -102,72 +101,88 @@ static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
  * TODO: docs
  */
 void *controller_run(void *arg) {
-
     controller_args_t *args = (controller_args_t *)(arg);
     controller_t controller;
     int err;
 
-    /* Connect to the controller */
-    err = controller_init(&controller, args->port);
-    if (err) {
-        fprintf(stderr, "Could not initialize controller with error: %s\n", strerror(err));
-        thread_return(err);
-    }
-    pthread_cleanup_push(controller_cleanup, &controller);
+    fprintf(stderr, "Waiting for controller...\n");
 
-    err = controller_accept(&controller);
-    if (err) {
-        fprintf(stderr, "Could not connect to controller with error: %s\n", strerror(err));
-        thread_return(err);
-    }
-    printf("Controller connected!\n");
-
-    /* Receive messages */
+    pthread_cleanup_push(controller_cleanup, &controller);
```

Comment:
Cleanup should only be pushed once the controller is actually initialized. Otherwise this creates a race condition where if the thread gets terminated before reaching `controller_init` we are trying to cleanup something that doesn't exist.

Reply:
I should also mention that if anywhere in the loop you already call `controller_disconnect` before `pthread_cleanup_pop` it will create the same race condition.

Reply:
> Cleanup should only be pushed once the controller is actually initialized. Otherwise this creates a race condition where if the thread gets terminated before reaching `controller_init` we are trying to cleanup something that doesn't exist.

Will putting pthread_cleanup inside the for loop cause issues if I push multiple clean up jobs? If it does, can I alternatively instead of calling disconnect call pthread_cleanup_pop(1)?

Reply:
Putting it inside the loop will cause issues likely yeah, it'll be pushing and popping on every iteration. You can manually call disconnect yourself if it's easier yes!